pkg/engines/jenkins/controller.go

```diff
@@ -0,0 +1,456 @@
+	"github.com/jenkins-x/lighthouse/pkg/config"
+	"github.com/jenkins-x/lighthouse/pkg/config/lighthouse"
+	"github.com/jenkins-x/lighthouse/pkg/jobutil"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/bwmarrin/snowflake"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+type lighthouseJobClient interface {
+	Create(*v1alpha1.LighthouseJob) (*v1alpha1.LighthouseJob, error)
+	List(metav1.ListOptions) (*v1alpha1.LighthouseJobList, error)
+	UpdateStatus(*v1alpha1.LighthouseJob) (*v1alpha1.LighthouseJob, error)
+	Patch(name string, pt types.PatchType, data []byte, subresources ...string) (result *v1alpha1.LighthouseJob, err error)
+}
+
+type jenkinsClient interface {
+	Build(*v1alpha1.LighthouseJob, string) error
+	ListBuilds(jobs []BuildQueryParams) (map[string]Build, error)
+	Abort(job string, build *Build) error
+}
+
+type syncFn func(v1alpha1.LighthouseJob, map[string]Build) error
+
+// Controller manages LighthouseJobs on a Jenkins server.
+type Controller struct {
+	lighthouseClient lighthouseJobClient
+	jenkinsClient    jenkinsClient
+	log              *logrus.Entry
+	cfg              config.Getter
+	node             *snowflake.Node
+	// if skip report job results to github
+	skipReport bool
+	// selector that will be applied on Lighthouse jobs.
+	selector string
+
+	lock sync.RWMutex
+	// pendingJobs is a short-lived cache that helps in limiting
+	// the maximum concurrency of jobs.
+	pendingJobs map[string]int
+
+	jobLock sync.RWMutex
+	// shared across the controller and a goroutine that gathers metrics.
+	jobs  []v1alpha1.LighthouseJob
+	clock clock.Clock
+}
+
+// NewController creates a new Controller from the provided clients.
+func NewController(lighthouseClient client.LighthouseJobInterface, jenkinsClient *Client, logger *logrus.Entry, cfg config.Getter, selector string) (*Controller, error) {
+	n, err := snowflake.NewNode(1)
+	if err != nil {
+		return nil, err
+	}
+	if logger == nil {
+		logger = logrus.NewEntry(logrus.StandardLogger())
+	}
+	return &Controller{
+		lighthouseClient: lighthouseClient,
+		jenkinsClient:    jenkinsClient,
+		log:              logger,
+		cfg:              cfg,
+		selector:         selector,
+		node:             n,
+		pendingJobs:      make(map[string]int),
+		clock:            clock.RealClock{},
+	}, nil
+}
+
+func (c *Controller) config() lighthouse.Controller {
+	operators := c.cfg().Jenkinses
+	if len(operators) == 1 {
+		return operators[0].Controller
+	}
+	configured := make([]string, 0, len(operators))
+	for _, cfg := range operators {
+		if cfg.LabelSelectorString == c.selector {
+			return cfg.Controller
+		}
+		configured = append(configured, cfg.LabelSelectorString)
+	}
+	if len(c.selector) == 0 {
+		c.log.Panicf("You need to specify a non-empty --label-selector (existing selectors: %v).", configured)
+	} else {
+		c.log.Panicf("No config exists for --label-selector=%s.", c.selector)
+	}
```

Comment:
I'm curious about this part. It looks like no matter `c.selector` is what it always panic.